### PR TITLE
[SPARK-49558][SQL] Add SQL pipe syntax for LIMIT/OFFSET and ORDER/SORT/CLUSTER/DISTRIBUTE BY

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

This PR adds SQL pipe syntax support for LIMIT/OFFSET and ORDER/SORT/CLUSTER/DISTRIBUTE BY.

For example:

```
CREATE TABLE t(x INT, y STRING) USING CSV;
INSERT INTO t VALUES (0, 'abc'), (1, 'def');

TABLE t
|> ORDER BY x
|> LIMIT 1 OFFSET 1

1	def
```

### Why are the changes needed?

The SQL pipe operator syntax will let users compose queries in a more flexible fashion.

### Does this PR introduce _any_ user-facing change?

Yes, see above.

### How was this patch tested?

This PR adds a few unit test cases, but mostly relies on golden file test coverage. I did this to make sure the answers are correct as this feature is implemented and also so we can look at the analyzer output plans to ensure they look right as well.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by dtenedor** [comment]
cc @cloud-fan @gengliangwang this is the PR to support LIMIT/OFFSET + sorting. There are a few more changes in the `AstBuilder` for this one but still contained only in the parser.

File: sql/core/src/test/resources/sql-tests/inputs/pipe-operators.sql [hunk 1]
```diff
@@ -571,6 +583,95 @@ table t
...
+|> order by x desc
+order by y;
+
+-- Sorting and repartitioning operators: negative tests.
+--------------------------------------------------------
+
+-- Multiple order by clauses are not supported in the same pipe operator
+table t
+|> order by x desc order by x + y
+order by y;
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/pipe-operators.sql:640]: Do we need this final order by to trigger the error?

File: sql/core/src/test/resources/sql-tests/results/pipe-operators.sql.out [hunk 2]
```diff
@@ -1673,6 +1691,279 @@ org.apache.spark.sql.catalyst.ExtendedAnalysisException
...
+    "fragment" : "x"
+  } ]
+}
+
+
+-- !query
+table t
+|> where x = 1
+|> select y
+|> offset 1
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/pipe-operators.sql.out:1867]: this doesn't fail?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 3]
```diff
@@ -1010,21 +1018,40 @@ class AstBuilder extends DataTypeAstBuilder
...
     }
 
     // WINDOWS
-    val withWindow = withOrder.optionalMap(windowClause)(withWindowClause)
+    val withWindow = withOrder.optionalMap(windowClause) {
+      withWindowClause
+    }
+    if (forPipeOperators && windowClause != null) {
+      throw QueryParsingErrors.combinationQueryResultClausesUnsupportedError(
+        ctx, s"WINDOW clauses within SQL pipe operators")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:1031]: This doesn't really match `COMBINATION_QUERY_RESULT_CLAUSES`, shall we add a new error condition for it?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 4]
```diff
@@ -1010,21 +1018,40 @@ class AstBuilder extends DataTypeAstBuilder
...
+    }
 
     // OFFSET
     // - OFFSET 0 is the same as omitting the OFFSET clause
+    val offsetClause = "OFFSET"
     val withOffset = withWindow.optional(offset) {
+      if (forPipeOperators && clause.nonEmpty) {
+        throw QueryParsingErrors.combinationQueryResultClausesUnsupportedError(
+          ctx, s"the $clause and $offsetClause clauses")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:1040]: shall we highlight it's only a limitation in the SQL pipe syntax?

File: sql/core/src/test/resources/sql-tests/results/pipe-operators.sql.out [hunk 5]
```diff
@@ -1673,6 +1691,279 @@ org.apache.spark.sql.catalyst.ExtendedAnalysisException
...
+|> order by x sort by x
+-- !query schema
+struct<>
+-- !query output
+org.apache.spark.sql.catalyst.parser.ParseException
+{
+  "errorClass" : "UNSUPPORTED_FEATURE.COMBINATION_QUERY_RESULT_CLAUSES",
+  "sqlState" : "0A000",
+  "messageParameters" : {
+    "clauses" : "ORDER BY/SORT BY/DISTRIBUTE BY/CLUSTER BY"
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/pipe-operators.sql.out:1908]: To match this existing error message, shall we change the pipe one to be `ORDER BY/LIMIT within pipe operator`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 3 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Sounds good, done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 4 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:1044]: Sounds good, done.

File: sql/core/src/test/resources/sql-tests/results/pipe-operators.sql.out (hunk 5 above)
Comment (dtenedor) [reply sql/core/src/test/resources/sql-tests/results/pipe-operators.sql.out:1908]: Sounds good, done.

**Review by dtenedor — COMMENTED** [review]
Thanks @cloud-fan for your review!

**Comment by dtenedor** [comment]
The most recent CI failure is not related

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 6]
```diff
@@ -1010,21 +1018,44 @@ class AstBuilder extends DataTypeAstBuilder
       // [EMPTY]
       query
     } else {
-      throw QueryParsingErrors.combinationQueryResultClausesUnsupportedError(ctx)
+      val allClauses = "ORDER BY/SORT BY/DISTRIBUTE BY/CLUSTER BY"
+      if (forPipeOperators) {
+        throw QueryParsingErrors.clausesWithPipeOperatorsUnsupportedError(ctx, allClauses)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:1023]: We should always call `combinationQueryResultClausesUnsupportedError` here, as this is not related to pipe. We don't support this combination in both pipe and classic SQL.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 7]
```diff
@@ -1010,21 +1018,44 @@ class AstBuilder extends DataTypeAstBuilder
...
       Offset(typedVisit(offset), withWindow)
     }
 
     // LIMIT
     // - LIMIT ALL is the same as omitting the LIMIT clause
     withOffset.optional(limit) {
+      val limitClause = "LIMIT"
+      if (forPipeOperators && clause.nonEmpty && clause != offsetClause) {
+        throw QueryParsingErrors.clausesWithPipeOperatorsUnsupportedError(
+          ctx, s"the $clause and $limitClause clauses")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:1056]:
```suggestion
          ctx, s"the co-existence of $clause and $limitClause clauses")
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 8]
```diff
@@ -1010,21 +1018,44 @@ class AstBuilder extends DataTypeAstBuilder
...
+        ctx, s"the WINDOW clause")
+    }
 
     // OFFSET
     // - OFFSET 0 is the same as omitting the OFFSET clause
+    val offsetClause = "OFFSET"
     val withOffset = withWindow.optional(offset) {
+      if (forPipeOperators && clause.nonEmpty) {
+        throw QueryParsingErrors.clausesWithPipeOperatorsUnsupportedError(
+          ctx, s"the $clause and $offsetClause clauses")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:1044]:
```suggestion
          ctx, s"the co-existence of $clause and $offsetClause clauses")
```

File: common/utils/src/main/resources/error/error-conditions.json [hunk 9]
```diff
@@ -4926,9 +4926,14 @@
...
         ]
       },
+      "CLAUSE_WITH_PIPE_OPERATORS" : {
+        "message" : [
+          "The SQL pipe operator syntax using |> does not support <clauses>."
+        ]
+      },
       "COMBINATION_QUERY_RESULT_CLAUSES" : {
         "message" : [
-          "Combination of ORDER BY/SORT BY/DISTRIBUTE BY/CLUSTER BY."
```
Comment (cloud-fan) [inline common/utils/src/main/resources/error/error-conditions.json:4931]: looking at the current code change, seems we don't need to update it anymore?

File: common/utils/src/main/resources/error/error-conditions.json [hunk 10]
```diff
@@ -4926,9 +4926,14 @@
           "Catalog <catalogName> does not support <operation>."
         ]
       },
+      "CLAUSE_WITH_PIPE_OPERATORS" : {
```
Comment (cloud-fan) [inline common/utils/src/main/resources/error/error-conditions.json:4929]: Maybe @srielau can also comment here, but it looks to me that we have two new errors:
1. WINDOW in the SQL pipe operator. We don't support window definition in SQL pipe yet.
2. More than one QUERY_RESULT_CLAUSES in SQL pipe. SQL pipe is designed to specify one operator at a time, so we don't allow any combination, such as ORDER BY col LIMIT 1, or LIMIT 1 OFFSET 1

Comment (srielau) [reply common/utils/src/main/resources/error/error-conditions.json:4929]: 1. appears to be a 0A000 (unsupported feature)
2. Isn't that a syntax error?

Comment (cloud-fan) [reply common/utils/src/main/resources/error/error-conditions.json:4929]: yes, 2 is a syntax error for the pipe statement.

Comment (dtenedor) [reply common/utils/src/main/resources/error/error-conditions.json:4929]: Thanks, split this into two new errors per suggestion.

**Review by dtenedor — COMMENTED** [review]
Thanks @cloud-fan for your reviews!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 6 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Sounds good, this is done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 7 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Sounds good, this is done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 8 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Sounds good, this is done.

File: common/utils/src/main/resources/error/error-conditions.json (hunk 9 above)
Comment (dtenedor) [reply common/utils/src/main/resources/error/error-conditions.json:4931]: Sounds good, this is done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 11]
```diff
@@ -1014,17 +1022,35 @@ class AstBuilder extends DataTypeAstBuilder
...
+      withWindowClause
+    }
+    if (forPipeOperators && windowClause != null) {
+      throw QueryParsingErrors.clausesWithPipeOperatorsUnsupportedError(
+        ctx, s"the WINDOW clause")
+    }
 
     // OFFSET
     // - OFFSET 0 is the same as omitting the OFFSET clause
+    val offsetClause = "OFFSET"
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:1035]: nit: shall we put all the constant clauses into an object?

**Review by gengliangwang — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 11 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Sure, it is good to keep magic strings out of the code.

File: common/utils/src/main/resources/error/error-conditions.json [hunk 12]
```diff
@@ -5006,6 +5011,11 @@
           "Multiple bucket TRANSFORMs."
         ]
       },
+      "MULTIPLE_QUERY_RESULT_CLAUSES_WITH_PIPE_OPERATORS" : {
```
Comment (cloud-fan) [inline common/utils/src/main/resources/error/error-conditions.json:5014]: shall we turn it into a top-level error class? This is syntax error, not unsupported feature, and should have a different SQL state.

File: common/utils/src/main/resources/error/error-conditions.json [hunk 13]
```diff
@@ -5006,6 +5011,11 @@
...
+        "message" : [
+          "Syntax error: the SQL pipe operator syntax using |> does not support <clauses>. Please separate the multiple result clauses into separate pipe operators and then retry the query again."
```
Comment (cloud-fan) [inline common/utils/src/main/resources/error/error-conditions.json:5016]: how about
```
<clause1> and <clause2> can not coexist in SQL pipe operator (1>). Please separate ...
```

Comment (cloud-fan) [reply common/utils/src/main/resources/error/error-conditions.json:5016]: The idea is to make the error parameters simple, so that people can consume it programatically.

File: common/utils/src/main/resources/error/error-conditions.json (hunk 12 above)
Comment (dtenedor) [reply common/utils/src/main/resources/error/error-conditions.json]: Sounds good, done.

File: common/utils/src/main/resources/error/error-conditions.json (hunk 13 above)
Comment (dtenedor) [reply common/utils/src/main/resources/error/error-conditions.json]: Sure, switched to this format.

**Review by dtenedor — COMMENTED** [review]
Thanks @cloud-fan again for your reviews!

File: common/utils/src/main/resources/error/error-conditions.json [hunk 14]
```diff
@@ -3448,6 +3448,11 @@
     ],
     "sqlState" : "0A000"
   },
+  "MULTIPLE_QUERY_RESULT_CLAUSES_WITH_PIPE_OPERATORS" : {
+    "message" : [
+      "<clause1> and <clause2> can not coexist in the same SQL pipe operator using '|>'. Please separate the multiple result clauses into separate pipe operators and then retry the query again."
+    ]
```
Comment (cloud-fan) [inline common/utils/src/main/resources/error/error-conditions.json:3454]: we missed the sqlState

**Review by cloud-fan — APPROVED** [review]

File: common/utils/src/main/resources/error/error-conditions.json (hunk 14 above)
Comment (dtenedor) [reply common/utils/src/main/resources/error/error-conditions.json]: 👍 I added the sqlState in here.

**Comment by cloud-fan** [comment]
The linter failure is unrelated, I'm merging it to master, thanks!

File: sql/core/src/test/resources/sql-tests/inputs/pipe-operators.sql [hunk 15]
```diff
@@ -571,6 +583,97 @@ table t
...
+table t
+|> order by x limit 1;
+
+-- ORDER BY and SORT BY are not supported at the same time.
+table t
+|> order by x sort by x;
+
+-- The WINDOW clause is not supported yet.
+table windowTestData
+|> window w as (partition by cte order by val)
```
Comment (Angryrou) [inline sql/core/src/test/resources/sql-tests/inputs/pipe-operators.sql:670]: Hi Daniel @dtenedor , I noticed that this window clause in Spark differs from what’s described in the original paper and documentation. Could you share your thoughts on this?

The [documentation](https://github.com/google/zetasql/blob/master/docs/pipe-syntax.md#window-pipe-operato
) specifies that a window operator should always include a window function with an OVER clause. However, in Spark's syntax, the window operator only returns a window definition without requiring an OVER clause.

I think it makes sense to keep the existing window syntax (as shown in this example) since the Extend clause will cover the window operator’s functionality as described in the paper. However, I’d like to confirm the expected behavior of the window clause in Spark SQL before proceeding with a PR.

Thanks in advance!

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/inputs/pipe-operators.sql:670]: In Spark we provide a default window frame so the OVER clause becomes optional. We shouldn't change it.
